sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedCollection.java

```diff
@@ -0,0 +1,74 @@
+package com.azure.android.core.util.paging;
+
+import androidx.annotation.NonNull;
+
+import java.util.Iterator;
+import java.util.List;
+
+public class PagedCollection<T, P extends Page<T>> implements Iterable<T> {
+    private static final int DEFAULT_PAGE_SIZE = 25; // Is this a good size?
+
+    PageRetriever<T, P> pageRetriever;
+
+    public PagedCollection(PageRetriever<T, P> pageRetriever) {
+        this.pageRetriever = pageRetriever;
+    }
+
+    public void nextPage(String pageId, int pageSize, PagingCallback<T, P> callback) {
+        pageRetriever.getPage(pageId, pageSize, callback);
+    }
+
+    // Not sure sync method should be removed.
+    public P nextPage(String pageId, int pageSize) {
+        return pageRetriever.getPage(pageId, pageSize);
+    }
+
+    @NonNull
+    @Override
+    public Iterator<T> iterator() {
+        return new Iterator<T>() {
+            private boolean hasNext;
+            private int currentItem;
+            private List<T> items;
+            private String itNextPageId;
+
+            @Override
+            public boolean hasNext() {
+                if (items == null) {
+                    updateItems();
+                }
+
+                return hasNext;
+            }
```

Comment:
Some quick feedback:

1. If there is a chance for an exception (e.g. failed to load a page) then it'll be good to throw from one place, say `hasNext`. 
If `hasNext` returns true i.e. without any exception then `next()` must return a value and should not throw. This means we should move all page loading logic to one place i.e. in `hasNext ()`.

2. We should stop fetching more page if "PageRetriever.getPage::onSuccess" give `null` value for "nextPageId", we can treat this as a contract.

3. We could use `items.iterator()` and could cut down some member variables, something like: 

```java

private Iterator<T> itemsItr;
private String nextPageId;

public boolean hasNext() {
    if (itemsItr == null) {
        loadNextPage();
    }
    boolean hasNext = itemsItr.hasNext();
    if (hasNext) {
        return true;
    } else {
        if (nextPageId == null) {
            return false;
        }
        loadNextPage();
        return itemsItr.hasNext();
    }
}

public T next() {
    return itemsItr.next();
}


private void loadNextPage() {
    retriever.getPage(..) {
        public void onSuccess(P page, String currentPageId, String nextPageId) {
            this.nextPageId = nextPageId;
            this.itemsItr = page.getIems() == null ? (new ArrayList<>).iterator() : page.getIems().iterator();
        }
    }
}
```

Reply:
Thank you for the feedback Anu, those are things I hadn't considered the first time around :)